src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,215 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (size == 0) {
+            newNode.prev = null;
+            newNode.next = null;
+            head = newNode;
+            tail = newNode;
+            incrementSize();
+        } else {
+            tail.next = newNode;
+            newNode.next = null;
+            newNode.prev = tail;
+            tail = newNode;
+            incrementSize();
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index != size) {
+            checkIndex(index);
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> currentNode = head;
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == 0) {
+            addInStart(newNode);
+            return;
+        } else {
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+            Node<T> prevNode = currentNode.prev;
+            newNode.next = currentNode;
+            newNode.prev = prevNode;
+            prevNode.next = newNode;
+            currentNode.prev = newNode;
+            incrementSize();
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        Node<T> currentNode = head;
+        if (size == 1) {
+            newNode.prev = null;
+            newNode.next = null;
+            head = newNode;
+            tail = newNode;
+            return currentNode.value;
+        }
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        T lastValue = currentNode.value;
+        currentNode.value = value;
+        return lastValue;
```

Comment:
to make this code more concise and readable, use the method `getNodeByIndex` and just reassign `value` of the returned node, it'll be 4 lines of code